src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionBatchFixAllProvider.cs

```diff
@@ -86,73 +86,61 @@ internal abstract class AbstractSuppressionBatchFixAllProvider : FixAllProvider
+                    }
+                    finally
+                    {
+                        args.progressTracker.ItemCompleted();
+                    }
+                },
+                consumeItems: static async (stream, args, cancellationToken) =>
+                {
+                    await foreach (var tuple in stream)
+                        args.results.Add(tuple);
+                },
+                args: (@this: this, fixAllState, progressTracker, results),
+                cancellationToken).ConfigureAwait(false);
 
-            await Task.WhenAll(tasks).ConfigureAwait(false);
-        }
-
-        return [.. fixesBag];
-    }
-
-    private async Task AddDocumentFixesAsync(
-        Document document, ImmutableArray<Diagnostic> diagnostics,
-        ConcurrentBag<(Diagnostic diagnostic, CodeAction action)> fixes,
-        FixAllState fixAllState, IProgress<CodeAnalysisProgress> progressTracker, CancellationToken cancellationToken)
-    {
-        try
-        {
-            await this.AddDocumentFixesAsync(document, diagnostics, fixes, fixAllState, cancellationToken).ConfigureAwait(false);
-        }
-        finally
-        {
-            progressTracker.ItemCompleted();
+            return results.ToImmutableAndClear();
         }
     }
 
     protected virtual async Task AddDocumentFixesAsync(
         Document document, ImmutableArray<Diagnostic> diagnostics,
-        ConcurrentBag<(Diagnostic diagnostic, CodeAction action)> fixes,
+        Action<(Diagnostic diagnostic, CodeAction action)> onItemFound,
         FixAllState fixAllState, CancellationToken cancellationToken)
     {
         Debug.Assert(!diagnostics.IsDefault);
         cancellationToken.ThrowIfCancellationRequested();
 
-        var registerCodeFix = GetRegisterCodeFixAction(fixAllState, fixes);
-
-        var fixerTasks = new List<Task>();
-        foreach (var diagnostic in diagnostics)
-        {
-            cancellationToken.ThrowIfCancellationRequested();
-            fixerTasks.Add(Task.Run(() =>
+        var registerCodeFix = GetRegisterCodeFixAction(fixAllState, onItemFound);
+        await RoslynParallel.ForEachAsync(
+            source: diagnostics,
+            cancellationToken,
+            async (diagnostic, cancellationToken) =>
             {
                 var context = new CodeFixContext(document, diagnostic, registerCodeFix, cancellationToken);
 
                 // TODO: Wrap call to ComputeFixesAsync() below in IExtensionManager.PerformFunctionAsync() so that
                 // a buggy extension that throws can't bring down the host?
-                return fixAllState.Provider.RegisterCodeFixesAsync(context) ?? Task.CompletedTask;
-            }, cancellationToken));
-        }
-
-        await Task.WhenAll(fixerTasks).ConfigureAwait(false);
+                var task = fixAllState.Provider.RegisterCodeFixesAsync(context) ?? Task.CompletedTask;
```

Comment:
>fixAllState.Provider.RegisterCodeFixesAsync(context) [](http://example.com/codeflow?start=27&length=52)

nit, maybe:

if (fixAllState.Provider.RegisterCodeFixesAsync(context) is Task task)
    await task.ConfigureAwait(false);

Reply:
done.